### PR TITLE
Allow bypassing futuroscope entirely

### PR DESCRIPTION
**Pull request by ivoanjo**
As discussed in https://github.com/codegram/hyperclient/issues/102 there are cases where we don't really want the async behavior provided by futuroscope.

In @Talkdesk's case, we're using hyperclient with the [manticore](https://github.com/cheald/manticore) gem that already has its own worker thread pool.

While the proposed `NoPool` approach kind-of works, that approach is global. This means that you chose once for your entire system whether you have a pool or not, and that choice impacts all hyperclient instances (even communicating with different systems) and futuroscope users (unless they allow you to override the pool, which for instance hyperclient doesn't).

Since that is rather heavy-handed, instead this PR adds an `async_options` to the `EntryPoint`, which allows hyperclient uses to per-instance specify if they want requests to be asynchronous or not. By default, the previous async behavior is followed, but this can be changed easily.

In the future I hope to also expand the `async_options` so you can specify a specific pool, so that you can have different pools for different hyperclient instances.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -187,6 +187,23 @@ spline._delete
...
+```ruby
+Futuroscope.default_pool.min_workers = 10
+Futuroscope.default_pool.max_workers = 20
+```
+
+If you want to disable this behavior and have all requests performed synchronously, you can use the `async_options` to disable this behavior:
+
+```ruby
+api = Hyperclient.new('https://grape-with-roar.herokuapp.com/api') do |client|
+  client.async_options = {enabled: false}
```
Comment (dblock) [inline README.md]: Minor: make it Ruby pretty, add a space after `{` and before `}`.

File: lib/hyperclient/entry_point.rb [hunk 2]
```diff
@@ -39,6 +39,7 @@ class EntryPoint < Link
     def initialize(url, &_block)
       @link = { 'href' => url }
       @entry_point = self
+      @async_options = {enabled: true}
```
Comment (dblock) [inline lib/hyperclient/entry_point.rb]: `rubocop -a` will fix this too

File: lib/hyperclient/link.rb [hunk 3]
```diff
@@ -167,9 +167,17 @@ def _uri_template
 
     def http_method(method, body = nil)
       @resource = begin
-        response = Futuroscope::Future.new do
-          @entry_point.connection.run_request(method, _url, body, nil)
-        end
+        do_async = @entry_point.async_options[:enabled]
```
Comment (dblock) [inline lib/hyperclient/link.rb]: Minor, but this isn't used in more than one place, so I would just do `if @entry_point.async ...`.

**Comment by dblock** [comment]
First, thank you, this is great. I feel like this should just be `options` and there should be `options[:async]`, this way it would be more extensible?

**Comment by ivoanjo** [comment]
Thanks for the feedback! I'll make the changes you suggested asap :)

File: lib/hyperclient/link.rb (hunk 3 above)
Comment (ivoanjo) [inline lib/hyperclient/link.rb]: I've updated this to do `if @entry_point.options[:async]` directly. Would you prefer I add an explicit `async` getter allowing this code to do `if @entry_point.async` instead?

**Comment by ivoanjo** [comment]
@dblock care to re-review? :smile:

File: README.md [hunk 4]
```diff
@@ -187,6 +187,23 @@ spline._delete
...
+```ruby
+Futuroscope.default_pool.min_workers = 10
+Futuroscope.default_pool.max_workers = 20
+```
+
+If you want to disable this behavior and have all requests performed synchronously, you can use the client `options` to disable this behavior:
+
+```ruby
+api = Hyperclient.new('https://grape-with-roar.herokuapp.com/api') do |client|
+  client.options.merge!(async: false)
```
Comment (dblock) [inline README.md]: No need to change this, but in Ruby this is an antipattern. All you're doing is setting `async`, so `client.options[:async] = false` is a much clearer operation than `merge`.

File: lib/hyperclient/entry_point.rb [hunk 5]
```diff
@@ -112,6 +113,18 @@ def faraday_block=(value)
...
+    #
+    # value    - A Hash containing the client options. Use { async: false } to
+    # to disable the default behavior of performing requests asynchronously
+    # using futures.
+    attr_writer :options
+
+    # Public: Read options.
+    #
+    # Returns a Hash.
+    attr_reader :options
```
Comment (dblock) [inline lib/hyperclient/entry_point.rb]: These should be combined as `attr_accessor`, no?

**Comment by dblock** [comment]
Merging this, thanks! Made minor comments if you want to address them in a future PR feel free.

File: lib/hyperclient/entry_point.rb (hunk 5 above)
Comment (ivoanjo) [inline lib/hyperclient/entry_point.rb]: I separated them to have documentation on both, but it may not be useful enough to warrant that distinction.

**Pull request merged**